### PR TITLE
refactor: screen rewritten as functional component

### PR DESCRIPTION
**Pull request by alduzy**
## Description

This PR intents to change Screen component into a functional component.

### CODE REVIEW

**Review by kkafar — COMMENTED** [review]
Hey, great job 🎉 I've got few remarks and questions we need to answer before we can merge these changes.

The rest of changes look very promising, but I would ask you to test them within our `TestsExample` application in following scenarios:

1. using native-stack v5 (`createNativeStackNavigator` imported from `react-native-screens`
2. using native-stack v6 (`createNativeStackNavigator` imported from `@react-navigation/native-stack`

Please be aware of prop naming differences between v5 & v6. 

We need to check few selected tests (1072, 1069 and some with reanimated should be amongst them) whether we don't have any undesired behaviour in downstream projects.

File: src/components/Screen.tsx [hunk 1]
```diff
@@ -42,27 +41,24 @@ interface ViewConfig extends View {
...
-  private progress = new Animated.Value(0);
-  private goingForward = new Animated.Value(0);
+export const InnerScreen = React.forwardRef<View, ScreenProps>(
+  function InnerScreen(props, ref) {
+    const innerRef = React.useRef<ViewConfig | null>(null);
+    React.useImperativeHandle(ref, () => innerRef.current!, []);
 
-  setNativeProps(props: ScreenProps): void {
-    this.ref?.setNativeProps(props);
-  }
```
[the remark this answers is not in the document]
Comment (kkafar) [reply src/components/Screen.tsx:53]: Or is this intentional & you are utilising this method defined on `innerRef`?

File: src/components/Screen.tsx [hunk 2]
```diff
@@ -42,27 +41,24 @@ interface ViewConfig extends View {
...
+    const setRef = (ref: ViewConfig) => {
+      innerRef.current = ref;
+    };
 
-  setRef = (ref: React.ElementRef<typeof View> | null): void => {
-    this.ref = ref;
-    this.props.onComponentRef?.(ref);
```
Comment (kkafar) [inline src/components/Screen.tsx:57]: I would like to see some comment why do we resign from calling this callback here

File: src/components/Screen.tsx [hunk 3]
```diff
@@ -42,27 +41,24 @@ interface ViewConfig extends View {
...
-  };
+    const closing = new Animated.Value(0);
+    const progress = new Animated.Value(0);
+    const goingForward = new Animated.Value(0);
```
[the remark this answers is not in the document]
Comment (alduzy) [reply src/components/Screen.tsx:56]: Maybe leveraging the useRef hook to store the values would be a good idea? What do you think?

[the remark this answers is not in the document]
Comment (kkafar) [reply src/components/Screen.tsx:56]: That was also my initial thought & I think we should go for this. 

In any case I would like to hear from someone who might now some internals (already tagged).

[the remark this answers is not in the document]
Comment (WoLewicki) [reply src/components/Screen.tsx:56]: Yeah, I can see that in the example they use `useRef` for it: https://reactnative.dev/docs/animated#example. If it works correctly then it seems like the way to go. I can also see that a similar concept is used in `react-navigation`: https://github.com/react-navigation/react-navigation/blob/d0abdee67f5db8cf39112af535846ffededfb21d/packages/react-native-tab-view/src/useAnimatedValue.tsx#L4

File: src/components/Screen.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (WoLewicki) [reply src/components/Screen.tsx:53]: I believe `Animated` still uses `setNativeProps` under the hood so I'd leave it. After quick talk with @piaskowyk, `react-native-reanimated` is not utilizing it, but still we would want it for `Animated`. @alduzy did you maybe check examples with `Animated` to see if they work when this code is removed? Would be also good to check with both `useNativeDriver` set to `false` and `true` and in `TestsExample` and `FabricTestsExample`.

File: src/components/Screen.tsx [hunk 4]
```diff
@@ -42,27 +41,25 @@ interface ViewConfig extends View {
...
 
-export class InnerScreen extends React.Component<ScreenProps> {
-  private ref: React.ElementRef<typeof View> | null = null;
-  private closing = new Animated.Value(0);
-  private progress = new Animated.Value(0);
-  private goingForward = new Animated.Value(0);
+export const InnerScreen = React.forwardRef<View, ScreenProps>(
+  function InnerScreen(props, ref) {
+    const innerRef = React.useRef<ViewConfig | null>(null);
+    React.useImperativeHandle(ref, () => innerRef.current!, []);
```
Comment (WoLewicki) [inline src/components/Screen.tsx:47]: What is this line for?

**Review by WoLewicki — COMMENTED** [review]
I answered some questions and left my own. Please answer them. Good job overall :tada:

File: src/components/Screen.tsx (hunk 4 above)
Comment (alduzy) [reply src/components/Screen.tsx:47]: This code is intended to replace removed code mentioned above:
`  setNativeProps(props: ScreenProps): void {
    this.ref?.setNativeProps(props);
  }`
  However I am not sure wether it is correctly implemented

**Review by WoLewicki — APPROVED** [review]
LGTM! If it has been tested with both `react-native-reanimated` and `Animated` and on both new and old architecture, then we can merge it 🚀

**Review by kkafar — APPROVED** [review]
+1 on the above comment from @WoLewicki

**Review by tboba — APPROVED** [review]
LGTM! 🥳

**Pull request merged**